code/Qt_VHDL_SVG_proj/doc/man/man3/SvgParser.3
.TH "SvgParser" 3 "VHDL simulator" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SvgParser \- The \fBSvgParser\fP class\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <SvgParser\&.h>\fP
.PP
Inherited by \fBSvgParseManager\fP\&.
.SS "Classes"

.in +1c
.ti -1c
.RI "struct \fBs_io\fP"
.br
.RI "Struct for storing input/output information\&. "
.ti -1c
.RI "struct \fBs_tree_node\fP"
.br
.RI "Struct for representing tree nodes\&. "
.ti -1c
.RI "struct \fBs_tree_node_info\fP"
.br
.RI "Struct for storing tree node information\&. "
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSvgParser\fP (\fBQString\fP \fBsvg_file\fP)"
.br
.RI "Constructor for \fBSvgParser\fP class Initializes the parsing process by loading the SVG file, generating a tree structure, and extracting component information\&. "
.ti -1c
.RI "\fB~SvgParser\fP ()"
.br
.RI "\fBSvgParser\fP destructor\&. Cleans up resources used by the parser\&. "
.ti -1c
.RI "\fBconst\fP \fBs_components_list\fP & \fBget_components_list\fP () \fBconst\fP"
.br
.in -1c
.SS "Protected Types"

.in +1c
.ti -1c
.RI "enum \fBe_types\fP { \fBlabel\fP, \fBtype\fP, \fBgraphic\fP, \fBinput\fP, \fBoutput\fP, \fBinputs\fP, \fBoutputs\fP, \fBsel\fP, \fBwire\fP }"
.br
.RI "Enumeration for the names of the type attribute\&. "
.ti -1c
.RI "enum \fBe_returns\fP { \fBSUCCESS\fP, \fBFILE_NOT_FOUND\fP, \fBINVALID_ARGUMENT\fP }"
.br
.RI "Enumeration for return codes\&. "
.ti -1c
.RI "enum \fBe_states\fP { \fBTO_BE_FOUND\fP }"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "\fBint\fP \fBgroups_number\fP"
.br
.RI "Number of groups\&. "
.ti -1c
.RI "\fBs_tree_node\fP \fBroot\fP"
.br
.RI "Root node of the tree\&. "
.ti -1c
.RI "\fBs_components_list\fP \fBall_components\fP"
.br
.RI "All components private variable\&. "
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "\fBQString\fP \fBattr\fP (\fBe_types\fP \fBtype\fP)"
.br
.RI "Gets the custom attribute name for a specific component type\&. "
.ti -1c
.RI "\fBQString\fP \fBattr_name_for_str\fP (\fBQString\fP \fBattr\fP)"
.br
.RI "Constructs a custom attribute name\&. "
.ti -1c
.RI "\fBvoid\fP \fBparse_one_element\fP (\fBconst\fP \fBQDomElement\fP \fBsvg_group_xml\fP, \fBs_elements\fP &elements)"
.br
.RI "Parses a single SVG element and extracts information related to the device type, name, label, inputs, and outputs\&. Adds the parsed element to the provided list of elements\&. "
.ti -1c
.RI "\fBvoid\fP \fBparse_element\fP (\fBs_tree_node\fP &\fBnode\fP)"
.br
.RI "Parses an individual element in the SVG file\&. "
.ti -1c
.RI "\fBvoid\fP \fBlist_attributes\fP (\fBconst\fP \fBQList\fP< \fBQDomElement\fP > &elements)"
.br
.RI "List attributes\&. "
.ti -1c
.RI "\fBvoid\fP \fBparse_svg\fP (\fBQString\fP \fBsvg_file\fP)"
.br
.RI "Parse SVG file\&. Opens the SVG file and processes its content, parsing the file into groups and building a tree structure of the elements\&. "
.ti -1c
.RI "\fBvoid\fP \fBparse_by_group\fP (\fBconst\fP \fBQDomNode\fP &\fBnode\fP, \fBs_tree_node\fP &\fBparentNode\fP, \fBint\fP level)"
.br
.RI "Recursively parses the SVG document by group\&. "
.ti -1c
.RI "\fBvoid\fP \fBget_group_header\fP (\fBconst\fP \fBQDomElement\fP &element, \fBs_tree_node_info\fP &\fBinfo\fP)"
.br
.RI "Extracts group header information from a DOM element\&. "
.ti -1c
.RI "\fBvoid\fP \fBgenerate_tree\fP (\fBconst\fP \fBs_tree_node\fP &\fBnode\fP, \fBconst\fP \fBQString\fP &\fBprefix\fP)"
.br
.RI "Generates a tree structure from parsed SVG groups\&. "
.ti -1c
.RI "\fBvoid\fP \fBprint_tree_in_log\fP (\fBQString\fP \fBlog_buffer\fP)"
.br
.RI "Prints the generated tree structure to the log\&. Outputs the parsed tree structure of SVG elements to the log window\&. "
.ti -1c
.RI "\fBvoid\fP \fBparse_components\fP (\fBs_tree_node\fP &\fBnode\fP)"
.br
.RI "Parses the components of the SVG file\&. Recursively processes the tree nodes and extracts component information from each element\&. "
.ti -1c
.RI "\fBvoid\fP \fBparse_simulation_IOs\fP (\fBconst\fP \fBQDomElement\fP \fBsvg_group_xml\fP, \fBs_sim_I_Os\fP &\fBparsed_IOs\fP)"
.br
.RI "Parse simulation IO\&. Parses simulation inputs and outputs (IOs) from the given SVG element\&. Adds the parsed IOs to the provided list, distinguishing between inputs and outputs based on the attributes in the XML\&. "
.ti -1c
.RI "\fBint\fP \fBget_list_of_outputs_name_and_width\fP (\fBQString\fP \fBoutputs_string\fP, \fBQList\fP< \fBs_element_io\fP > &\fBoutputs_list\fP)"
.br
.RI "Parse outputs from string Parses a list of output names and their corresponding widths from a string formatted as 'name:width,name:width'\&. Adds the parsed outputs to the provided list\&. "
.ti -1c
.RI "\fBint\fP \fBget_list_of_inputs_name_and_width\fP (\fBQString\fP \fBoutputs_string\fP, \fBQList\fP< \fBs_element_io\fP > &\fBinputs_list\fP)"
.br
.RI "Parse inputs from string Parses a list of input names, their corresponding widths, and connected components from a string formatted as 'name:width:connected_to,name:width:connected_to'\&. Adds the parsed inputs to the provided list\&. "
.ti -1c
.RI "\fBbool\fP \fBcheck_and_get_attr\fP (\fBconst\fP \fBQDomElement\fP &\fBxml\fP, \fBQString\fP &\fBstr_to_get\fP, \fBQString\fP \fBattr_name\fP)"
.br
.RI "Get attribute value if it exist Checks if the specified XML element contains the given attribute and, if found, retrieves the attribute's value\&. "
.ti -1c
.RI "\fBvoid\fP \fBparse_simulation_wires\fP (\fBconst\fP \fBQDomElement\fP \fBsvg_group_xml\fP, \fBs_sim_wires\fP &\fBparsed_wires\fP)"
.br
.RI "Parse simulation wires\&. Parses the wires in the simulation from the given SVG element, extracting connection and name information\&. Adds the parsed wires to the provided structure\&. "
.ti -1c
.RI "\fBvoid\fP \fBlist_matching_attr_with_value\fP (\fBconst\fP \fBQDomElement\fP \fBelem_to_look_into\fP, \fBconst\fP \fBQString\fP \fBattr_name\fP, \fBconst\fP \fBQString\fP \fBattr_value\fP, \fBQList\fP< \fBQDomElement\fP > &\fBfound_elements\fP)"
.br
.RI "Find elements with attribute\&. Searches the given XML element for child elements that contain a specific attribute with a matching value\&. Adds any matching elements to the provided list\&. "
.ti -1c
.RI "\fBvoid\fP \fBadd_error_message\fP (\fBs_parse_error\fP &error, \fBconst\fP \fBQString\fP &\fBerrorMessage\fP)"
.br
.RI "Attach an error message to a structure\&. Adds an error message to the given error structure\&. Marks the structure as having a parse error\&. "
.ti -1c
.RI "\fBint\fP \fBlist_matching_attr\fP (\fBconst\fP \fBQDomElement\fP \fBelem_to_look_into\fP, \fBconst\fP \fBQString\fP \fBattr_name\fP, \fBQList\fP< \fBQDomElement\fP > &\fBfound_elements\fP)"
.br
.RI "list matching attribute Searches the given XML element for child elements that contain a specific attribute\&. Adds any matching elements to the provided list\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBQString\fP \fBsvg_file\fP"
.br
.RI "File name of the SVG file\&. "
.ti -1c
.RI "\fBconst\fP \fBQString\fP \fBcustom_attribute\fP = \fBCUSTOM_ATTRIBUTE\fP"
.br
.RI "Custom attribute prefix\&. "
.ti -1c
.RI "\fBDebugWindow\fP * \fBdebug\fP"
.br
.ti -1c
.RI "\fBQString\fP \fBlog_buffer\fP"
.br
.RI "Buffer for log messages\&. "
.in -1c
.SH "Detailed Description"
.PP 
The \fBSvgParser\fP class\&. 

This class parses an SVG file and extracts electrical links from attributes of SVG elements\&. 
.SH "Member Enumeration Documentation"
.PP 
.SS "\fBenum\fP \fBSvgParser::e_returns\fP\fC [protected]\fP"

.PP
Enumeration for return codes\&. 
.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fISUCCESS \fP\fP
Success return code\&. 
.TP
\fB\fIFILE_NOT_FOUND \fP\fP
File not found return code\&. 
.TP
\fB\fIINVALID_ARGUMENT \fP\fP
Invalid argument return code\&. 
.SS "\fBenum\fP \fBSvgParser::e_states\fP\fC [protected]\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fITO_BE_FOUND \fP\fP
To be found state\&. 
.SS "\fBenum\fP \fBSvgParser::e_types\fP\fC [protected]\fP"

.PP
Enumeration for the names of the type attribute\&. 
.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIlabel \fP\fP
Label attribute type\&. 
.TP
\fB\fItype \fP\fP
Type attribute type\&. 
.TP
\fB\fIgraphic \fP\fP
Graphic attribute type\&. 
.TP
\fB\fIinput \fP\fP
Input attribute type\&. 
.TP
\fB\fIoutput \fP\fP
Output attribute type\&. 
.TP
\fB\fIinputs \fP\fP
Inputs attribute type\&. 
.TP
\fB\fIoutputs \fP\fP
Outputs attribute type\&. 
.TP
\fB\fIsel \fP\fP
Selection attribute type\&. 
.TP
\fB\fIwire \fP\fP
Wire attribute type\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SvgParser::SvgParser (\fBQString\fP svg_file)"

.PP
Constructor for \fBSvgParser\fP class Initializes the parsing process by loading the SVG file, generating a tree structure, and extracting component information\&. 
.PP
\fBParameters\fP
.RS 4
\fIsvg_file\fP The path to the SVG file to be parsed\&. 
.RE
.PP

.SS "SvgParser::~SvgParser ()"

.PP
\fBSvgParser\fP destructor\&. Cleans up resources used by the parser\&. 
.SH "Member Function Documentation"
.PP 
.SS "\fBvoid\fP SvgParser::add_error_message (\fBs_parse_error\fP & error, \fBconst\fP \fBQString\fP & errorMessage)\fC [private]\fP"

.PP
Attach an error message to a structure\&. Adds an error message to the given error structure\&. Marks the structure as having a parse error\&. 
.PP
\fBParameters\fP
.RS 4
\fIerror\fP The structure where the error message will be added\&. 
.br
\fIerrorMessage\fP The error message to be added\&. 
.RE
.PP

.SS "\fBQString\fP SvgParser::attr (\fBe_types\fP type)\fC [private]\fP"

.PP
Gets the custom attribute name for a specific component type\&. Returns the attribute name corresponding to the provided component type\&.
.PP
\fBParameters\fP
.RS 4
\fItype\fP The type of the component\&. 
.RE
.PP
\fBReturns\fP
.RS 4
The custom attribute name for the given component type\&. 
.RE
.PP

.SS "\fBQString\fP SvgParser::attr_name_for_str (\fBQString\fP attr)\fC [private]\fP"

.PP
Constructs a custom attribute name\&. Generates a custom attribute name by prepending a defined namespace to the input string\&.
.PP
\fBParameters\fP
.RS 4
\fIattr\fP The input attribute name\&. 
.RE
.PP
\fBReturns\fP
.RS 4
The full custom attribute name\&. 
.RE
.PP

.SS "\fBbool\fP SvgParser::check_and_get_attr (\fBconst\fP \fBQDomElement\fP & xml, \fBQString\fP & str_to_get, \fBQString\fP attr_name)\fC [private]\fP"

.PP
Get attribute value if it exist Checks if the specified XML element contains the given attribute and, if found, retrieves the attribute's value\&. 
.PP
\fBParameters\fP
.RS 4
\fIxml\fP The QDomElement to check\&. 
.br
\fIstr_to_get\fP Reference to the QString where the attribute value will be stored\&. 
.br
\fIattr_name\fP The name of the attribute to check for\&. 
.RE
.PP
\fBReturns\fP
.RS 4
True if the attribute is found, false otherwise\&. 
.RE
.PP

.SS "\fBvoid\fP SvgParser::generate_tree (\fBconst\fP \fBs_tree_node\fP & node, \fBconst\fP \fBQString\fP & prefix)\fC [private]\fP"

.PP
Generates a tree structure from parsed SVG groups\&. Outputs a visual representation of the tree structure of SVG elements, including their tag names and IDs\&.
.PP
\fBParameters\fP
.RS 4
\fInode\fP The current node in the tree structure\&. 
.br
\fIprefix\fP The prefix used to format the tree structure display\&. 
.RE
.PP

.SS "\fBconst\fP \fBs_components_list\fP & SvgParser::get_components_list () const\fC [inline]\fP"

.SS "\fBvoid\fP SvgParser::get_group_header (\fBconst\fP \fBQDomElement\fP & element, \fBs_tree_node_info\fP & info)\fC [private]\fP"

.PP
Extracts group header information from a DOM element\&. Parses a group element (<g>) in the SVG file to retrieve its attributes (\fBe\&.g\fP\&., device type, width)\&.
.PP
\fBParameters\fP
.RS 4
\fIelement\fP The current DOM element being processed\&. 
.br

code/Qt_VHDL_SVG_proj/parser_linker/SvgParser.cpp
#include "SvgParser.h"
#include "debugwindow.h"

SvgParser::SvgParser(QString svg_file)
{
    log_buffer = "";
    root.level = 0;

    debug = DebugWindow::getInstance();


    if (svg_file.isEmpty())
    {
        debug->addError("Empty file name");

        return;
    }

    this->svg_file = svg_file;
    parse_svg(this->svg_file);

    // Get the tree of the elements in the SVG file
    generate_tree(root, "  ");
    print_tree_in_log(log_buffer);

    // Parse the components info and links from the SVG file (wire, input, output, modules, etc.)
    parse_components(root);
}

SvgParser::~SvgParser()
{
    // Destructor
}

void SvgParser::parse_svg(QString svg_file)
{
    // Open the SVG file
    QFile file(svg_file);
    QDomDocument doc;
    try
    {
        if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
        {
            throw std::runtime_error("Failed to open SVG file");
        }
        if (!doc.setContent(&file))
        {
            debug->addError("Failed to parse SVG file as XML.");
        }
    }
    catch (const std::exception &e)
    {
        debug->addError(e.what());
        return;
    }

    // Create the root node of the tree and get the main SVG element infos
    root.element = doc.documentElement();
    get_group_header(root.element, root.infos);

    // 1.Parse the SVG in different groups where <g> is the parent element and contains an ID
    groups_number = 0;
    parse_by_group(root.element, root, 0);

    // Output found groups
    DebugWindow::getInstance()->addDebug("Found " + QString::number(groups_number) + " groups:");

    file.close();
}

void SvgParser::parse_by_group(const QDomNode &node, s_tree_node &parentNode, int level)
{ // Recursively find all groups nodes in a node and return a vector of QDomElements
    QDomNodeList children = node.childNodes();
    for (int i = 0; i < children.count(); ++i)
    {
        QDomNode childNode = children.at(i);
        if (childNode.isElement())
        {
            QDomElement childElement = childNode.toElement();
            if (!childElement.isNull() && childElement.tagName() == "g")
            {
                ++groups_number;
                s_tree_node childTreeNode;
                childTreeNode.element = childElement;
                get_group_header(childElement, childTreeNode.infos);
                childTreeNode.level = level + 1; // Increment the level
                parentNode.children.push_back(childTreeNode);
                parse_by_group(childElement, parentNode.children.back(), level + 1);
            }
        }
    }
}

void SvgParser::generate_tree(const s_tree_node &node, const QString &prefix)
{
    QString adjusted_prefix = prefix;
    QString log_content; // Accumulate log output

    if (!prefix.isEmpty())
    {
        log_content += prefix;
        log_content += "|";
    }
    log_content += "- " + node.infos.tag_name;

    if (!node.infos.id.isEmpty())
    {
        log_content += " (id: " + node.infos.id + ")";
    }
    log_content += " - Level: " + QString::number(node.level);

    // Accumulate log output into a buffer
    log_buffer += log_content + "\n";

    for (int i = 0; i < node.children.size(); ++i)
    {
        if (i == node.children.size() - 1)
        {
            adjusted_prefix += "   "; // Adjust prefix for the last child
        }
        else
        {
            adjusted_prefix += "|  "; // Adjust prefix for other children
        }
        generate_tree(node.children[i], adjusted_prefix);
        adjusted_prefix = prefix; // Reset prefix for next child
    }
}

void SvgParser::print_tree_in_log(QString buffer)
{
    debug->addInfo( "SVG Groups Tree:");
    QStringList lines = buffer.split("\n", Qt::SkipEmptyParts);

    // Print log content from first line to last line
    for (int i = 0; i < lines.size(); ++i)
    {
        debug->addInfo(lines[i]);
    }

    debug->addInfo("End of SVG Groups Tree");
    // reset the log buffer
    buffer = "";
}

void SvgParser::parse_components(s_tree_node &node)
{
    // Parse the current node's element
    parse_element(node);

    debug->addInfo("Parsing links for node: " + node.infos.type + " at level " + QString::number(node.level));

    // Recursively call parse_tree_nodes()
    for (s_tree_node &child : node.children)
    {
        parse_components(child);
    }
}

QString SvgParser::attr_name_for_str(QString attr)
{
    return custom_attribute + ":" + attr;
}

QString SvgParser::attr(e_types type)
{
    switch (type)
    {
    case e_types::label:
        return custom_attribute + ":label";
        break;
    case e_types::type:
        return custom_attribute + ":type";
        break;
    case e_types::graphic:
        return custom_attribute + ":graphic";
        break;
    case e_types::input:
        return custom_attribute + ":input";
        break;
    case e_types::output:
        return custom_attribute + ":output";
        break;
    case e_types::inputs:
        return custom_attribute + ":inputs";
        break;
    case e_types::outputs:
        return custom_attribute + ":outputs";
        break;
    case e_types::sel:
        return custom_attribute + ":sel";
        break;
    case e_types::wire:
        return custom_attribute + ":wire";
        break;
    default:
        return "";
        break;
    }
}

void SvgParser::get_group_header(const QDomElement &element, s_tree_node_info &infos)
{
    // We should consistently have those infos

    // For the device type(MUX, AND, OR, etc)
    if (element.hasAttribute(attr_name_for_str("device")))
    {
        infos.type = element.attribute(attr_name_for_str("device"));
    }
    else
    {
        infos.type = "";
        infos.is_parse_error = true;
        infos.error_messages.push_back("No device attribute found");
    }

    // For the i/o width
    if (element.hasAttribute(attr_name_for_str("width")))
    {
        QString str_width = element.attribute(attr_name_for_str("width"));
        // Width is in the form "x,y" where x is the inputs width and y the outputs width
        QStringList widths = str_width.split(",");
        if (widths.size() == 2)
        {
            infos.component_in_width = widths[0].toInt();
            infos.component_out_width = widths[1].toInt();
        }
        else
        {
            infos.component_in_width = 0;
            infos.component_out_width = 0;
            infos.is_parse_error = true;
            infos.error_messages.push_back("Invalid width attribute format");
        }
    }
    else
    {
        infos.component_in_width = 0;
        infos.component_out_width = 0;
        infos.is_parse_error = true;
        infos.error_messages.push_back("No width attribute found");
    }
}

void SvgParser::parse_element(s_tree_node &node)
{
    QDomElement node_xml_data = node.element;    // Store the group element svg data
    s_tree_node_info *parsed_data = &node.infos; // Store parsed data here

    parsed_data->is_parse_error = false;
    parsed_data->has_been_parsed = false;

    if (node_xml_data.tagName() == "svg")
    { // Main element, contains groups, wires, inputs and outputs
        // We should only have one svg element
        if (parsed_data->has_been_parsed)
        {
            debug->addError("Main element parsed again");
            return;
        }

        // Parse the simulator inputs and outputs
        parse_simulation_IOs(node_xml_data, all_components.simulation_IOs);

        // Parse the wires
        parse_simulation_wires(node_xml_data, all_components.simulation_wires); // TODO : Wire parsing not working


        // No header for the main element, hard code the values
        parsed_data->type = "main";
        parsed_data->component_in_width = 0;
        parsed_data->component_out_width = 0;

        parsed_data->has_been_parsed = true;
    }
    else if (node_xml_data.tagName() == "g")
    {
        // Group element, contains components
        parse_one_element(node_xml_data, all_components.elements);
    }
    else
    {
#ifdef DEBUG
        debug->addError("Unknown element type");
#endif
    }
}

void SvgParser::parse_one_element(const QDomElement svg_group_xml, s_elements &elements)
{
    /**
     * On an element, we should have the following attributes:
     * - device: the type of the component
     * - label : the schematic name of the component
     * - inputs: the list of inputs -> d0:1:IN0,d1:1:IN1   (d0:1:IN0 means input 0, width 1, connected to IN0)
     * - outputs: the list of outputs -> d0:1,d1:1
     */
    s_element new_element;
    new_element.error.is_parse_error = false;
    new_element.inputs_number = 0;

    // Get the device type
    new_element.device = svg_group_xml.attribute(ATTR_FOR_STR("device"));
    if (new_element.device.isEmpty())
    {
        add_error_message(new_element.error, "No device attribute found");
        new_element.device = "unknown";
    }

    // Get the name
    new_element.name = svg_group_xml.attribute(NAME_ATTRIBUTE);
    if (new_element.name.isEmpty())
    {
        add_error_message(new_element.error, "No name attribute found");
        new_element.name = "unknown";
    }

    // Get the label
    //It is a text contained in the group with the attribute sim:type="label"
    QList<QDomElement> label_list;
    list_matching_attr_with_value(svg_group_xml, attr(type), "label", label_list);
    if(label_list.size() == 1)
    {
        new_element.label = label_list.at(0).text();
    }
    else
    {
        add_error_message(new_element.error, "No or invalid label attribute");
        new_element.label = "unknown";
    }

    // Get the inputs and outputs
    QString str_inputs = svg_group_xml.attribute(ATTR_FOR_STR("inputs"));
    get_list_of_inputs_name_and_width(str_inputs, new_element.inputs);
    new_element.inputs_number = new_element.inputs.count();

    QString str_outputs = svg_group_xml.attribute(ATTR_FOR_STR("outputs"));
    get_list_of_outputs_name_and_width(str_outputs, new_element.outputs);
    new_element.outputs_number = new_element.outputs.count();

#ifdef DEBUG
    debug->addDebug("Device:" + new_element.device);
    debug->addDebug("Name:" + new_element.name);
    debug->addDebug("Label:" + new_element.label);
    for (auto &input : new_element.inputs)
    {
        debug->addDebug("Input: " + input.name +"connected to " + input.connected_to + "width:" + QString::number(input.width));
    }
    for (auto &output : new_element.outputs)
    {
        debug->addDebug("Output:" + output.name + "width:" + QString::number(output.width));
    }

    if(new_element.error.is_parse_error)
    {
        for(auto &errorMessage:new_element.error.error_messages)
        {
            debug->addError(errorMessage);
        }
    }
    else
    {
        debug->addSuccess(new_element.name + " parsed");
    }

#endif

    elements.elements_list.push_back(new_element);
}

void SvgParser::parse_simulation_IOs(const QDomElement svg_group_xml, s_sim_I_Os &parsed_IOs)
{
    QList<s_sim_I_O> sim_IOs;

    // Get both inputs and outputs in a list
    QList<QDomElement> inputs_and_outputs_found;
    list_matching_attr_with_value(svg_group_xml, attr(type), "sim_input", inputs_and_outputs_found);
    list_matching_attr_with_value(svg_group_xml, attr(type), "sim_output", inputs_and_outputs_found);

    parsed_IOs.error.is_parse_error = false;

    for (const QDomElement &element : inputs_and_outputs_found)
    {
        // Create a new sim_IO object to populate with the data from the xml element
        s_sim_I_O sim_IO;

        sim_IO.error.error_messages = {};
        sim_IO.error.is_parse_error = false;

        // Get if input or output and the connections name and width
        QString type;
        if (check_and_get_attr(element, type, ATTR_FOR_STR("type")))
        {
            if (type.compare("sim_input") == 0)
            { // If the element is an input
                sim_IO.type = INPUT;
                // It should have an attribute "sim:outputs" that contains the name of the output -> name:width, name:width,...
                //  output because the sim input outputs to the circuit
                QString outputs_string = "";
                if (check_and_get_attr(element, outputs_string, attr(outputs)))
                { // If the attribute exists
                    QList<s_element_io> outputs_list;
                    if (get_list_of_outputs_name_and_width(outputs_string, outputs_list) == 1)
                    { // If the attribute is valid (Only one output with correct syntax)
                        if (outputs_list[0].name.compare(FLAG_NAME_FROM_INKSCAPE) == 0)
                        { // If the name of the output is FLAG_NAME_FROM_INKSCAPE, the name will be the inkscape name
                            if (!check_and_get_attr(element, sim_IO.name, NAME_ATTRIBUTE))
                            { // If the name attribute does not exist
                                add_error_message(sim_IO.error, "Trying to use inkscape name but no name attribute found");
                                sim_IO.name = DEFAULT_INPUT_NAME + QString::number(sim_IOs.size());
                            }
                        }
                        else
                        { // Else, the name will be the one in the attribute
                            sim_IO.name = outputs_list[0].name;
                        }
                        sim_IO.width = outputs_list[0].width;
                    }
                    else
                    { // If the attribute is not valid
                        add_error_message(sim_IO.error, "Invalid outputs attribute number of outputs");
                        sim_IO.width = DEFAULT_WIDTH;
                        sim_IO.name = DEFAULT_INPUT_NAME + QString::number(sim_IOs.size());
                    }
                }
                else
                { // If the attribute does not exist
                    sim_IO.width = DEFAULT_WIDTH;
                    add_error_message(sim_IO.error, "No output attribute found for input");
                    sim_IO.name = DEFAULT_INPUT_NAME + QString::number(sim_IOs.size());
                }
            }
            else if (type.compare("sim_output") == 0)
            { // If the element is a simulator output
                sim_IO.type = OUTPUT;

                // It should have an attribute "inputs" that contains the name,width and the name of the output symbol it is connected to
                // Input because the sim output takes the value from the circuit as an input
                QString inputs_string = "";
                if (check_and_get_attr(element, inputs_string, attr(inputs)))
                { // If the attribute exists
                    QList<s_element_io> inputs_list;

                    if (get_list_of_inputs_name_and_width(inputs_string, inputs_list) == 1)
                    { // If the attribute is valid (Only one input with correct syntax)
                        if (inputs_list[0].name.compare(FLAG_NAME_FROM_INKSCAPE) == 0)
                        { // If the name of the input is FLAG_NAME_FROM_INKSCAPE, the name will be the inkscape name
                            if (!check_and_get_attr(element, sim_IO.name, NAME_ATTRIBUTE))
                            { // If the name attribute does not exist
                                add_error_message(sim_IO.error, "Trying to use inkscape name but no name attribute found");
                                sim_IO.name = DEFAULT_INPUT_NAME + QString::number(sim_IOs.size());
                            }
                        }
                        else
                        { // Else, the name will be the one in the attribute
                            sim_IO.name = inputs_list[0].name;
                        }
                        sim_IO.width = inputs_list[0].width;
                        sim_IO.connected_to = inputs_list[0].connected_to;
                    }
                    else
                    { // If the attribute is not valid
                        add_error_message(sim_IO.error, "Invalid inputs attribute number of inputs");
                        sim_IO.width = DEFAULT_WIDTH;
                        sim_IO.name = DEFAULT_OUTPUT_NAME + QString::number(sim_IOs.size());
                        sim_IO.connected_to = "";
                    }
                }
                else
                { // If the attribute does not exist
                    sim_IO.width = DEFAULT_WIDTH;
                    add_error_message(sim_IO.error, "No inputs attribute found for output");
                    sim_IO.name = DEFAULT_OUTPUT_NAME + QString::number(sim_IOs.size());
                    sim_IO.connected_to = "";
                }
            }
            else
            {
                sim_IO.type = UNDEFINED;
                add_error_message(sim_IO.error, "No type attribute found for an undifined IO");
                continue; // Skip this element
            }
        }
        else
        {
            sim_IO.type = UNDEFINED;
            add_error_message(sim_IO.error, "No attribute found for an IO");
            continue; // Skip this element
        }
        parsed_IOs.i_os.push_back(sim_IO);
        parsed_IOs.error.is_parse_error |= sim_IO.error.is_parse_error;
    }
}

int SvgParser::get_list_of_outputs_name_and_width(QString outputs_string, QList<s_element_io> &outputs_list)
{
    // Outputs should be formated as : name:width,name:width
    QStringList outputs = outputs_string.split(",");
    for (QString &output : outputs)
    {
        s_element_io new_elem_output;

        output = output.trimmed(); // Remove white spaces
        QStringList output_data = output.split(":");
        if (output_data.size() == 2)
        {
            // TODO : verify that the width is a number
            new_elem_output.name = output_data[0].trimmed();
            new_elem_output.width = output_data[1].toInt();
            outputs_list.append(new_elem_output);
        }
        else
        {
            return FORMAT_ERROR;
        }
    }
    // Return the number of outputs found
    return outputs.size();
}

int SvgParser::get_list_of_inputs_name_and_width(QString inputs_string, QList<s_element_io> &inputs_list)
{
    // Inputs should be formated as : name:width:connected_to,name:width:connected_to,...
    QStringList inputs = inputs_string.split(",");
    for (QString &input : inputs)
    {
        s_element_io new_elem_IO;

        input = input.trimmed(); // Remove white spaces
        QStringList output_data = input.split(":");

        if (output_data.size() == 3)
        {//Make sure that the input is formated well
            new_elem_IO.name = output_data[0].trimmed();
            new_elem_IO.width = output_data[1].toInt();
            new_elem_IO.connected_to = output_data[2].trimmed();
            inputs_list.append(new_elem_IO);
        }
        else
        {
            return FORMAT_ERROR;
        }
    }
    // Return the number of inputs found
    return inputs.size();
}

bool SvgParser::check_and_get_attr(const QDomElement &xml, QString &str_to_get, QString attr_name)
{
    if (xml.hasAttribute(attr_name))
    {
        str_to_get = xml.attribute(attr_name);
        return true;
    }
    else
    {
        return false;
    }
}

void SvgParser::add_error_message(s_parse_error &error, const QString &errorMessage)
{
    error.is_parse_error = true;
    error.error_messages.push_back(errorMessage);
}

void SvgParser::parse_simulation_wires(const QDomElement svg_group_xml, s_sim_wires &parsed_wires)
{
    QList<s_sim_wire> sim_wires;
    s_sim_wire sim_wire;
    sim_wire.error.is_parse_error = false;
    sim_wire.error.error_messages = {};

    bool error_on_wires = false;

    QList<QDomElement> wires_found;

    // Extract the wires from the xml of the <svg> element
    list_matching_attr_with_value(
        svg_group_xml,
        attr(type),
        "wire",
        wires_found);

    for (const QDomElement &wire_found : wires_found)
    {
        // Parse the name
        if (wire_found.hasAttribute(NAME_ATTRIBUTE))
        {
            sim_wire.name = wire_found.attribute(NAME_ATTRIBUTE);
        }
        else
        {
            QString error_message = "No name attribute found for wire ";
            sim_wire.name = DEFAULT_WIRE_NAME + QString::number(sim_wires.size());
            sim_wire.error.is_parse_error = true;
            sim_wire.error.error_messages.push_back(error_message);
        }

        // Parse the connection
        if (wire_found.hasAttribute(attr_name_for_str("connection")))
        {
            sim_wire.connected_to = wire_found.attribute(attr_name_for_str("connection"));
        }
        else
        {
            QString error_message = "No connection attribute found for wire ";
            sim_wire.connected_to = "UNKNOWN";
            sim_wire.error.is_parse_error = true;
            sim_wire.error.error_messages.push_back(error_message + sim_wire.name);
        }
        // The wire width is from the connection width we'll find that info later
        sim_wire.width = 0;

        // Get a global error state for the wires
        error_on_wires |= sim_wire.error.is_parse_error;

        parsed_wires.wires.append(sim_wire);
        parsed_wires.error.is_parse_error = error_on_wires;
    }
}

void SvgParser::list_matching_attr_with_value(const QDomElement elem_to_look_into, const QString attr_name, const QString attr_value, QList<QDomElement> &found_elements)
{
    // Find attributes with a fixed value
    QDomElement child = elem_to_look_into.firstChildElement();
    while (!child.isNull())
    {
        if (child.attributes().contains(attr_name))
        {
            if (child.attribute(attr_name) == attr_value)
            {
                // debug->add child.attribute(attr_name) << "==" << attr_value;
                found_elements.append(child);
            }
        }
        child = child.nextSiblingElement();
    }
}

int SvgParser::list_matching_attr(const QDomElement elem_to_look_into, const QString attr_name, QList<QDomElement> &found_elements)
{
    // Find attributes with a fixed value
    QDomElement child = elem_to_look_into.firstChildElement();
    while (!child.isNull())
    {
        if (child.attributes().contains(attr_name))
        {
            found_elements.append(child);
        }
        child = child.nextSiblingElement();
    }
    return found_elements.size();
}

